### PR TITLE
Remove packages folder from Compile and EmbeddedResource item lists

### LINKED ISSUE #306: Match Compile, EmbeddedResource project.json defaults
There's been an ask to clean up the compile and embedded resource items post-migration. 

To do this safely (without risk of mangling someones project), SDK will need to update it's default excludes to match that of project.json. 

These are the full set of defaults for project.json

```
<Compile Include="**\*.cs" Exclude="bin\**;obj\**;**\*.xproj;packages\**" />

<EmbeddedResource Include="**\*.resx" Exclude="bin\**;obj\**;**\*.xproj;packages\**" />
<EmbeddedResource Include="compiler\resources\**\*" Exclude="bin\**;obj\**;**\*.xproj;packages\**" />
```

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes #306

### CODE REVIEW

**Comment by dsplaisted** [comment]
Should we have any sort of testing for the default excludes?  I didn't find any.

Also, repeating my [comment on the issue](https://github.com/dotnet/sdk/issues/306#issuecomment-255491061): What about `node_modules` or similar folders?  I thought one of the requirements for decent globbing perf was to avoid looking at those folders, but I don't see them excluded anywhere.

**Comment by nguerrera** [comment]
Tests would definitely be good. ;)

**Review by nguerrera — APPROVED** [review]

**Comment by nguerrera** [comment]
node_modules makes sense. Any other package manager conventions we should add?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: could we add **\*.xproj as well? This would match exactly what we currently have in project.json world.

Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: @livarcocc The only case where adding `**\*.xproj` here would have an effect is if there is an `.xproj` in under the `compiler\resources` folder.  Is there any normal scenario where that would happen?

**Comment by nguerrera** [comment]
Also xprojs will cease to exist, right? We aren't doing future maintainers a favor if they end up staring at a default about format that never hit 1.0. If you want 100% guaranteed compat on that, then add the xproj exclude to the compiler/resources include in migration. I'd rather this relic live in migrated projects alone than in defaults for the rest of time.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: I see, so this does not get pulled into the excludes for compile and embedded resources? That's where we have this exclude in the CLI.

**Comment by livarcocc** [comment]
We have it in the migration, but I would like to remove it from there so that those projects are clean(er).

I was just working under the assumption from this issue here opened by Sri: https://github.com/dotnet/cli/issues/4330. Item 5 is about excludes.

**Comment by nguerrera** [comment]
I think having xproj mentioned in these targets is instant legacy. Won't folks want to delete them anyway? Maybe just leave them included on migration? Can we get data about how often this exclude actually triggers in practice?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: It does apply to those but the corresponding includes will be filtered by extension (*.cs, *.vb, *.resx). Those won't match xproj. The one exception is that project.json included compiler/resources/**/* (no extension filter) as resources. That's why it's the only case where the xproj exclude matters.

**Comment by srivatsn** [comment]
I agree we shouldn't mention xprojs in the exclude. Migration would convert and delete those files where applicable and so we shouldn't run into any cases that need it.

**Comment by livarcocc** [comment]
Ok. We will move the xprojs over to a backup folder and I WILL NOT add this exclude to the excludes during migration.

Given that these are people moving from project.json to csproj, I want to keep the csproj to a bare minimum of verbosity.

**Comment by dsplaisted** [comment]
I've added `node_modules` and `jspm_packages` per feedback from the web team, and added a way to opt out of removing these files if you do want to include something from any of these folders.

**Comment by eerhardt** [comment]
> I've added node_modules and jspm_packages per feedback from the web team

Why can't the web sdk add removes for these folders?  Why should the core .NET SDK care about npm or jspm?  Those are web project concerns.  The can put them in the web sdk nupkg, like they did in the folloiwng:  https://github.com/dotnet/sdk/pull/223/files

**Comment by srivatsn** [comment]
Agree with Eric. Also don't they already have excludes for content, they could add for Compile as well.

**Comment by dsplaisted** [comment]
OK, I've sent a PR to the web Sdk to remove the node and jspm package folders from the item lists.  What about the `packages` folder.  Should we remove that in the .NET Sdk, or in the Web Sdk?

**Comment by eerhardt** [comment]
> What about the packages folder. Should we remove that in the .NET Sdk, or in the Web Sdk?

Since `packages` is typically NuGet's folder, I'm fine with the .NET Sdk excluding it.  We already know/care about NuGet in the .NET Sdk.

**Comment by dsplaisted** [comment]
I've updated this based on feedback and added tests.  It should be ready for review now.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs
Comment (nguerrera) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: Why do these need to be special cased? Also, why are there two files?

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs
Comment (nguerrera) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: Same question, can you comment and factor out a helper?

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs
Comment (nguerrera) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: nit: extra line of whitespace

**Review by nguerrera — APPROVED** [review]

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: One is for the `TargetFrameworkAttribute` and comes from the temp folder.  The other is for the other assembly attributes that the SDK adds, and is in the intermediate output path.

I've factored this out into a separate method that removes these, with a comment to explain what's going on.

**Comment by dsplaisted** [comment]
@srivatsn @MattGertz for approval

#### Scenario 
Parity in behavior with project.json projects, which automatically ignored files in the packages folder.

#### Bug
#306

#### Workarounds
You could add a `Remove` or `Exclude` yourself in the project file.  But this would mean that `dotnet migrate` would have to do so which would make the migrated projects uglier than they need to be.

#### Risk
Low risk - this only applies to projects using the .NET SDK.

#### Performance Impact
If anything it should improve performance as if there's a packages folder it won't need to be scanned for files that match a glob.

#### Regression Analysis
n/a

**Comment by MattGertz** [comment]
@dsplaisted Looks good, once Sri signs off I will approve.

**Review by srivatsn — APPROVED** [review]

**Comment by MattGertz** [comment]
Approved.

**Pull request merged**
